### PR TITLE
Refactor code to reuse common structs

### PR DESCRIPTION
**Pull request by kyessenov**
- Move `ProxyContext` to top-level.
- Define an interface for `ServiceAccounts`
- Use `Options` to pass flags to controllers consistently.

Should be no functional change.

### CODE REVIEW

File: test/integration/app.yaml.tmpl [hunk 1]
```diff
@@ -47,6 +47,8 @@ spec:
           - "{{.port3}}"
           - --port
           - "{{.port4}}"
+          - --port
```
Comment (kyessenov) [inline test/integration/app.yaml.tmpl]: will be used for health checking, please ignore this for now.

File: proxy/envoy/config.go [hunk 2]
```diff
@@ -128,9 +120,10 @@ func buildListeners(context *ProxyContext) (Listeners, Clusters) {
 	outbound, outClusters := buildOutboundListeners(instances, services, context)
 
 	listeners := append(inbound, outbound...)
-	listeners.normalize()
+	clusters := append(inClusters, outClusters...)
 
-	clusters := append(inClusters, outClusters...).normalize()
+	listeners = listeners.normalize()
```
Comment (ayj) [inline proxy/envoy/config.go]: Looks like this code is unrelated to the refactoring?

Comment (kyessenov) [inline proxy/envoy/config.go]: Yes, but it's a better style to split two operations into two statements I think, so I left it in this PR. Code looks cleaner.

Comment (ayj) [inline proxy/envoy/config.go]: SGTM

**Comment by kyessenov** [comment]
jenkins build manager/presubmit

**Comment by kyessenov** [comment]
Jenkins build manager/e2e-smoketest
